File: src/Plotly.NET/Traces/ObjectAbstractions/Table.fs
namespace Plotly.NET.TraceObjects

open Plotly.NET
open Plotly.NET.LayoutObjects
open DynamicObj
open System
open System.Runtime.InteropServices

/// CellColor type inherits from dynamic object
type TableFill() =
    inherit DynamicObj()

    static member init(?Color: Color) =
        TableFill() |> TableFill.style (?Color = Color)

    static member style(?Color: Color) =
        fun (fill: TableFill) ->
            fill
            |> DynObj.withOptionalProperty "color" Color

/// Cells type inherits from dynamic object
type TableCells() =
    inherit DynamicObj()

    /// Initialized Cells object
    static member init
        (
            ?Align: StyleParam.HorizontalAlign,
            ?MultiAlign: seq<StyleParam.HorizontalAlign>,
            ?Fill: TableFill,
            ?Font: Font,
            ?Format: seq<string>,
            ?Height: int,
            ?Line: Line,
            ?Prefix: string,
            ?MultiPrefix: seq<string>,
            ?Suffix: string,
            ?MultiSuffix: seq<string>,
            ?Values: seq<#seq<#IConvertible>>
        ) =
        TableCells()
        |> TableCells.style (
            ?Align = Align,
            ?MultiAlign = MultiAlign,
            ?Fill = Fill,
            ?Font = Font,
            ?Format = Format,
            ?Height = Height,
            ?Line = Line,
            ?Prefix = Prefix,
            ?MultiPrefix = MultiPrefix,
            ?Suffix = Suffix,
            ?MultiSuffix = MultiSuffix,
            ?Values = Values
        )

    //Applies the styles to TableCells()
    static member style
        (
            ?Align: StyleParam.HorizontalAlign,
            ?MultiAlign: seq<StyleParam.HorizontalAlign>,
            ?Fill: TableFill,
            ?Font: Font,
            ?Format: seq<string>,
            ?Height: int,
            ?Line: Line,
            ?Prefix: string,
            ?MultiPrefix: seq<string>,
            ?Suffix: string,
            ?MultiSuffix: seq<string>,
            ?Values: seq<#seq<#IConvertible>>
        ) =
        fun (cells: TableCells) ->

            cells
            |> DynObj.withOptionalSingleOrMultiPropertyBy "align" (Align, MultiAlign) StyleParam.HorizontalAlign.convert
            |> DynObj.withOptionalProperty "fill" Fill
            |> DynObj.withOptionalProperty "font" Font
            |> DynObj.withOptionalProperty "format" Format
            |> DynObj.withOptionalProperty "height" Height
            |> DynObj.withOptionalProperty "line" Line
            |> DynObj.withOptionalSingleOrMultiProperty "prefix" (Prefix, MultiPrefix)
            |> DynObj.withOptionalSingleOrMultiProperty "suffix" (Suffix, MultiSuffix)
            |> DynObj.withOptionalProperty "values" Values

type TableHeader = TableCells

File: src/Plotly.NET.CSharp/ObjectAbstractions/TraceObjects/Table.cs

using Microsoft.FSharp.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Plotly.NET.CSharp;

public static partial class TraceObjects
{
    public static Plotly.NET.TraceObjects.TableFill InitTableFill
        (
            Optional<Color> Color = default
        ) 
            =>
                Plotly.NET.TraceObjects.TableFill.init(
                    Color: Color.ToOption()
                );
    public static Plotly.NET.TraceObjects.TableCells InitTableCells<ValuesType>
        (
            Optional<StyleParam.HorizontalAlign> Align = default,
            Optional<IEnumerable<StyleParam.HorizontalAlign>> MultiAlign = default,
            Optional<Plotly.NET.TraceObjects.TableFill> Fill = default,
            Optional<Font> Font = default,
            Optional<IEnumerable<string>> Format = default,
            Optional<int> Height = default,
            Optional<Line> Line = default,
            Optional<string> Prefix = default,
            Optional<IEnumerable<string>> MultiPrefix = default,
            Optional<string> Suffix = default,
            Optional<IEnumerable<string>> MultiSuffix = default,
            Optional<IEnumerable<IEnumerable<ValuesType>>> Values = default
        ) 
            where ValuesType: IConvertible
            =>
                Plotly.NET.TraceObjects.TableCells.init(
                    Align: Align.ToOption(),
                    MultiAlign: MultiAlign.ToOption(),
                    Fill: Fill.ToOption(),
                    Font: Font.ToOption(),
                    Format: Format.ToOption(),
                    Height: Height.ToOption(),
                    Line: Line.ToOption(),
                    Prefix: Prefix.ToOption(),
                    MultiPrefix: MultiPrefix.ToOption(),
                    Suffix: Suffix.ToOption(),
                    MultiSuffix: MultiSuffix.ToOption(),
                    Values: Values.ToOption()
                );
}

